File: solutions/go/0045-permutations.go
func permute(nums []int) [][]int {
    var result [][]int
    
    var backtrack func([]int, []int)
    backtrack = func(nums []int, path []int) {
        if len(nums) == 0 {
            result = append(result, append([]int(nil), path...))
            return
        }
        for i := 0; i < len(nums); i++ {
            newNums := append([]int(nil), nums[:i]...)
            newNums = append(newNums, nums[i+1:]...)
            newPath := append([]int(nil), path...)
            newPath = append(newPath, nums[i])
            backtrack(newNums, newPath)
        }
    }
    
    backtrack(nums, []int{})
    return result
}

File: solutions/rust/0045-permutations.rs
impl Solution {
    pub fn permute(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut result = Vec::new();
        Self::backtrack(nums, vec![], &mut result);
        result
    }

    fn backtrack(nums: Vec<i32>, path: Vec<i32>, result: &mut Vec<Vec<i32>>) {
        if nums.is_empty() {
            result.push(path);
            return;
        }
        for i in 0..nums.len() {
            let mut new_nums = nums.clone();
            new_nums.remove(i);
            let mut new_path = path.clone();
            new_path.push(nums[i]);
            Self::backtrack(new_nums, new_path, result);
        }
    }
}
